### DISCUSSION TITLE
Limit length for simple_cycles

### DISCUSSION THREAD

**Discussion by vakker** [networkx/networkx#4797] [Algorithms]
Would that be possible to include an optional parameter to `simple_cycles` that limits the length of the cycles?
In my application I only need the cycles up to a certain length, but currently I find all of them and just filter out the long ones, which is very wasteful. 

I assume that during the search for cycles it would be fairly straight forward to break out of the loop when a specific depth is reached, e.g. somewhere around [here](https://github.com/networkx/networkx/blob/main/networkx/algorithms/cycles.py#L205) something like:
```python
elif nextnode not in blocked and len(path) < max_len:
```

**Comment by dschult** [comment]
I think that because `simple_cycles` yields the cycles in order of length (I think), that you can do this already with code like (untested):

```python
cycle_list = []
for cycle in nx.simple_cycles(...):
    if len(cycle) > max_len:
        break
    # process the cycle
```
Does that accomplish what you need?

**Reply by vakker** [reply]
Thanks for the suggestion, but that's not guaranteed I think. E.g. the following script:
```python
import networkx as nx

G = nx.fast_gnp_random_graph(10, 0.5, directed=True)
for c in nx.simple_cycles(G):
    print(len(c))
```
prints:
```
9                                                   
7                                                   
10                                                  
8                                                   
8                                                   
9                                             
7                                
.......
4
4
3
2
3
5
3
4
2
2
2
2
```
or using this graph from the [example](https://networkx.org/documentation/networkx-1.11/reference/generated/networkx.algorithms.cycles.simple_cycles.html#simple-cycles): 
```python
G = nx.DiGraph([(0, 0), (0, 1), (0, 2), (1, 2), (2, 0), (2, 1), (2, 2)])
```
prints:
```
1                                                   
1                                                   
2                                                   
3                                                   
2    
```

**Comment by boothby** [comment]
Cycles are not constructed in length of order.

In [5]: for c in nx.simple_cycles(nx.DiGraph(nx.complete_graph(3))):
   ...:     print(c)

[0, 2, 1]
[0, 2]
[0, 1, 2]
[0, 1]
[1, 2]

This isn't the first such request we've had, and filtering cycles on length
takes significantly longer than it needs to.  I believe that it should be
possible to skip over cycles longer than a user is interested in, because
the enumeration algorithm is stack-based

**Comment by vakker** [comment]
What about this:
```python
@not_implemented_for("undirected")
def simple_cycles(G, max_len=None):
    # ....

    def _unblock(thisnode, blocked, B):
        stack = {thisnode}
        while stack:
            node = stack.pop()
            if node in blocked:
                blocked.remove(node)
                stack.update(B[node])
                B[node].clear()

    def check_len(path):
        if max_len:
            return len(path) < max_len
        return True

    # Johnson's algorithm requires some ordering of the nodes.
    # We assign the arbitrary ordering given by the strongly connected comps
    # There is no need to track the ordering as each node removed as processed.
    # Also we save the actual graph so we can mutate it. We only take the
    # edges because we do not want to copy edge and node attributes here.
    subG = type(G)(G.edges())
    sccs = [scc for scc in nx.strongly_connected_components(subG) if len(scc) > 1]

    # Johnson's algorithm exclude self cycle edges like (v, v)
    # To be backward compatible, we record those cycles in advance
    # and then remove from subG
    for v in subG:
        if subG.has_edge(v, v):
            yield [v]
            subG.remove_edge(v, v)

    while sccs:
        scc = sccs.pop()
        sccG = subG.subgraph(scc)
        # order of scc determines ordering of nodes
        startnode = scc.pop()
        # Processing node runs "circuit" routine from recursive version
        path = [startnode]
        blocked = set()  # vertex: blocked from search?
        closed = set()  # nodes involved in a cycle
        blocked.add(startnode)
        B = defaultdict(set)  # graph portions that yield no elementary circuit
        stack = [(startnode, list(sccG[startnode]))]  # sccG gives comp nbrs
        while stack:
            thisnode, nbrs = stack[-1]
            if nbrs and check_len(path): #there are neighbours and the length is good
                nextnode = nbrs.pop()
                if nextnode == startnode:
                    yield path[:]
                    closed.update(path)
                #                        print "Found a cycle", path, closed
                elif nextnode not in blocked:
                    path.append(nextnode)
                    stack.append((nextnode, list(sccG[nextnode])))
                    closed.discard(nextnode)
                    blocked.add(nextnode)
                    continue
            # done with nextnode... look for more neighbors
            else:  # no more nbrs
                if thisnode in closed:
                    _unblock(thisnode, blocked, B)
                else:
                    for nbr in sccG[thisnode]:
                        if thisnode not in B[nbr]:
                            B[nbr].add(thisnode)
                stack.pop()
                #                assert path[-1] == thisnode
                path.pop()
        # done processing this node
        H = subG.subgraph(scc)  # make smaller to avoid work in SCC routine
        sccs.extend(scc for scc in nx.strongly_connected_components(H) if len(scc) > 1)
```

Testing it with this:
```python
G = nx.DiGraph(nx.complete_graph(9))

print(len(list(nx.simple_cycles(G))))
print(len(list(nx.simple_cycles(G, 5))))
```
gives
```
125664
960
```

**Comment by boothby** [comment]
I'll need to read Johnson[1] to make sure I understand how this will interact with blocking / unblocking, but yeah, that's approximately the solution I came up with.  I'd like to see this feature; please submit a PR.

[1] https://www.cs.tufts.edu/comp/150GA/homeworks/hw1/Johnson%2075.PDF

**Comment by vakker** [comment]
See #4798

**Comment by boothby** [comment]
My apologies, but this modification does foul blocking / unblocking.  I wrote the following to mine for counterexamples,

```python
import random, networkx as nx 
while 1: 
    g = nx.DiGraph(nx.complete_graph(4)) 
    while g.number_of_edges(): 
        g.remove_edge(*random.choice(list(g.edges())) ) 
        cyc = {tuple(c) for c in nx.simple_cycles(g)} 
        for k in range(2, len(g)+1): 
            A = {c for c in cyc if len(c) < k}
            B = {tuple(c) for c in nx.simple_cycles(g, k)} 
            assert A == B
```

which uncovers a discrepancy between reported 2-cycles of `nx.DiGraph([(0, 1), (0, 2), (1, 0), (2, 1), (3, 1)])`, and 3-cycles of `nx.DiGraph([(0, 1), (0, 3), (1, 2), (2, 0), (3, 1)])`.  I haven't found a suitable algorithm in the literature yet, but I haven't spent much time looking.  Are you interested in digging deeper?

**Reply by vakker** [reply]
~~Good catch, I changed `if not nbrs` to `else` which is not correct. I pushed a fix that doesn't have the inconsistency, let me know what you think about it.~~

No, this still doesn't work, I'll keep looking.

**Comment by vakker** [comment]
So I looked into the algorithm a bit more and I don't understand some parts of the implementation in the library.
E.g. in this block:
```python
            if not nbrs:  # no more nbrs
                if thisnode in closed:
                    _unblock(thisnode, blocked, B)
                else:
                    for nbr in sccG[thisnode]:
                        if thisnode not in B[nbr]:
                            B[nbr].add(thisnode)
                stack.pop()
                path.pop()
```
the `else` part never gets called. I think that would be called if there's a dead end in the graph, but that's filtered out anyway by
```python
sccs = [
        scc for scc in nx.strongly_connected_components(subG) if len(scc) > 1
    ]
```
It seems like `closed` is not being used in the paper. But I might be wrong about this.

So anyway, this version (based on [this](https://stackoverflow.com/a/48591893)) seems to work:
```python
def simple_cycles(G, max_len=None):
    def _check_len(path):
        if max_len:
            return len(path) < max_len
        return True

    subG = type(G)(G.edges())
    sccs = [
        scc for scc in nx.strongly_connected_components(subG) if len(scc) > 1
    ]

    while sccs:
        scc = sccs.pop()
        startnode = scc.pop()
        path = [startnode]
        blocked = set()
        blocked.add(startnode)
        stack = [(startnode, list(subG[startnode]))]

        while stack:
            thisnode, nbrs = stack[-1]

            if nbrs and _check_len(path):
                nextnode = nbrs.pop()
                if nextnode == startnode:
                    yield path[:]
                elif nextnode not in blocked:
                    path.append(nextnode)
                    stack.append((nextnode, list(subG[nextnode])))
                    blocked.add(nextnode)
            else:
                blocked.remove(thisnode)
                stack.pop()
                path.pop()
        subG.remove_node(startnode)
        H = subG.subgraph(scc)
        sccs.extend(list(nx.strongly_connected_components(H)))

**Comment by boothby** [comment]
Thanks for your work on this!  After reading through `recursive_simple_cycles` and Johnson, I think I understand where `closed` comes from.  In Johnson, the variable `f` corresponds to the variable `closed` used in `recursive_simple_cycles`.  From there, it gets quite interesting.  In the non-recursive `simple_cycles`, the `closed` variable was not preserved on the stack, but made into a set.  I must admit, I do not fully understand the algorithmic transformation employed there.

The utility of `closed` is certainly non-obvious.  Indeed, the solution you've found (quite similar to what I've used in the past when I needed some/all short cycles) appears to be similar to the algorithm of Tarjan (https://ecommons.cornell.edu/handle/1813/5941) that Johnson cites.  Johnson's algorithm keeps nodes blocked for longer than Tarjan's, which prevents the algorithm from unnecessarily exploring portions of the graph which won't produce cycles.  That is, the `else` statement is evaluated on some graphs.  If you're curious, Johnson provides a family of graphs for which there should be a visible discrepancy between the runtime of the two approaches.

The code you've written above appears to be correct, but it changes the runtime when `max_len is None`.  I haven't given up hope that Johnson's algorithm can be appropriately modified, but I don't have a solution at hand.  I'll try and poke at it some this weekend.  Failing that, I'd recommend adding this function under the name `all_bounded_simple_cycles` or similar, and dispatch that when `max_len is not None`.  That way, we maintain runtime in the unbounded case, and it will be good enough in the bounded case.

**Comment by boothby** [comment]
I finally returned to this in #6461 -- it's now reasonably fast to search for cycles of bounded length using `nx.simple_cycles`.
